### PR TITLE
Add Timeline frame number display to Preview

### PR DESCRIPTION
**Pull request by SuslikV**
Requested: https://github.com/OpenShot/openshot-qt/issues/891#issuecomment-608693012

OpenShot frame_number

Increases convenience of _Export_ usage in case only range of frames are required for final video render (value can be copied).

Later it can be modified into "Go to: Frame Number".

### CODE REVIEW

**Comment by SuslikV** [comment]
OpenShot frame_number02

What to wait for? Navigation is added.

**Comment by ferdnyc** [comment]
Neat! I'll check this out now.

**Comment by ferdnyc** [comment]
Quick impressions from testing, haven't looked at the code yet.

#### Layout
Adding this control shifts the transport controls off-center to the left.

When @N3WWN added the snapshot button, an invisible button was inserted on the left side of the toolbar to balance out the centering. We may need to do the same here, though that's going to make the minimum toolbar size awfully wide.

Other option, it could go on the left **instead** of that invisible button, which would help balance out the toolbar better. (In fact we'd need an invisible spacer on the _right_ side, to make up the difference between this width and the snapshot button width.)

Left-aligning would also help prevent it getting pushed off of the toolbar so quickly, at narrow widths. Right now as you narrow the preview dock, first the snapshot button disappears, then the frame number field shortly after, even though there's actually a fair amount of space remaining. If we widen the spacer(s) on the left, that behavior's going to trigger even sooner.

#### Updating
The value doesn't reset when loading a new project. That appears to be a larger bug, because hitting Play _will_ start from the frame indicated. So, if fixing that bug fixes this display as well, then great. But we should fix that as part of this change, because the disparity between the playhead positioning and the actual start-of-playback position (which exists even without this change) becomes glaringly visible in the UI.

It also doesn't update when dragging the left/right edge of a clip (editing the clip's Start or End/Duration properties), while the preview frame does change. Perhaps during those operations, the frame# shown should be either:
1. The timeline frame# that corresponds to the current drag position
2. The start or end frame# within the clip (The clip frame# that's visible in preview, IOW.)

_Both_ feel like they could be useful things to have, so I'm not sure which is better.

#### Performance
It's at least worth considering whether updates should be suspended during playback, the way the Properties table and the timeline scroll position are. We've already got so much going on when the preview is playing.

Measuring relative impact on playback isn't an easy thing (especially since I don't have enough RAM to run two copies of OpenShot at once), but it doesn't appear to be very great. Still, the question is really: what does the user gain from updating during playback / what do they lose if it doesn't?

**Comment by ferdnyc** [comment]
Oh, this could also do with a keyboard shortcut to focus/activate it.

* <kbd>Ctrl</kbd>+<kbd>G</kbd> is taken, it toggles snapping.
* <kbd>Insert</kbd> and <kbd>Home</kbd> are both free, and seem like possibilities 
    (<kbd>Ctrl</kbd>+<kbd>Home</kbd> is jump-to-start, but bare <kbd>Home</kbd> isn't used)
* Other possibilities, that I believe are free: <kbd>.</kbd> / <kbd>Ctrl</kbd>+<kbd>.</kbd>

(When picking a shortcut, we should consider a possible future "Go to time" feature, and the shortcut we'd assign to activate that. So perhaps whatever we pick, it should be <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+(something), so that "Go to time" can be assigned to the unshifted version.)

**Comment by SuslikV** [comment]
> Left-aligning would also help prevent it getting pushed off of the toolbar so quickly, at narrow widths...

It looks ugly. I tried it first.

**More about design.**

As for the alignment - it is strange that you didn't noticed that Save Frame button is not at center as all other controls... Look at YouTube's player playback button - should it be at the center too? VLC Player has default skin where Playback button is on the left side too.

Should I move all controls to the center? For me it is logical that frame number displayed near to the Save Frame button.

Maybe UI layouts can be updated later in other PR to not mix design vs new features?

**Shortcuts**

> Oh, this could also do with a keyboard shortcut to focus/activate it...

This makes feature very complicated and hidden from casual users. Application is not aimed at pro level at all. I'm against this.


**Performance**

Preview updates only at the end of the navigation (0.5 after playback paused, if it is not paused at the exact frame - if this even possible...).

Are you afraid of Qt events overheat?


**Behavior**

> The value doesn't reset when loading a new project...

I can look at this. Probably, simple value update after the load ended can solves this.


** Needs Clarification**

> It also doesn't update when dragging the left/right edge of a clip...

This is Timeline frame display (at preview profile fps) not a Clip frame. I not quite sure what you what to say here.

**Comment by ferdnyc** [comment]
> **Needs Clarification**
> 
> > It also doesn't update when dragging the left/right edge of a clip...
> 
> This is Timeline frame display (at preview profile fps) not a Clip frame. I not quite sure what you what to say here.

Try dragging the edge of a clip. The _video_ preview frame changes. It follows the drag position, and previews only the clip being resized (instead of the entire timeline) — IOW, it becomes a clip preview temporarily, to allow clip start/end positions to be set precisely. So, it could be helpful to also temporarily show the frame# _local to the current clip_, during that drag operation (only).

**Comment by ferdnyc** [comment]
> **Shortcuts**
> 
> > Oh, this could also do with a keyboard shortcut to focus/activate it...
> 
> This makes feature very complicated and hidden from casual users. Application is not aimed at pro level at all. I'm against this.

How does adding a shortcut change anything for casual users? I'm not talking about hiding the widget or requiring a shortcut to activate it, users can still click on it if they prefer. The widget would work no differently if it _has_ a shortcut than how it works without one. The shortcut just provides an additional means of interacting with it, for users who prefer to do things that way.

**Comment by ferdnyc** [comment]
> It looks ugly. I tried it first.

Yeah, I figured it wouldn't look ideal, though that may just be a case of needing to get used to it.

But like I said, it can still be placed on the right. There just needs to be additional spacing added on the left to balance it.

> As for the alignment - it is strange that you didn't noticed that Save Frame button is not at center as all other controls...

Didn't notice...? I specifically _mentioned_ that the control avoids throwing off centering of the transport controls, and how.

It is possible to add controls that occupy one side of the toolbar _without_ de-centering the transport controls. When @N3WWN added Save Frame, he not only made sure to do that, he even _added instructions_ for anyone making similar changes in the future.

https://github.com/OpenShot/openshot-qt/blob/3f26703f2ee6d3992b7827447ab06b552ed111ee/src/windows/main_window.py#L2041-L2047

(I'm to blame for subsequently removing the commented-out second example. But it wouldn't have applied anyway, here, since your control is more than `32px` wide.)

> Look at YouTube's player playback button - should it be at the center too? VLC Player has default skin where Playback button is on the left side too.

Irrelevant. Those are both video players, applications where the Play button is the primary focus and all other controls are secondary. Controls for video _editing_ software are different.

Regardless, a new arrangement for the controls could be discussed, but _that_ is indeed a topic best left to a separate PR/project. For the moment, they are centered under the preview, and should be kept centered under the preview.

> Should I move all controls to the center?

No, of course not, I never even mentioned that.

> For me it is logical that frame number displayed near to the Save Frame button.

I can see that. Not sure it's really that strong a connection, but OK, there's some logic to that. And like I said, that's completely possible, just as long as the left is offset to preserve centering.

> Maybe UI layouts can be updated later in other PR to not mix design vs new features?

If it were possible to implement this feature without changing any design elements, sure. But it's not, and this PR affects the application design, so they were already "mixed" right from the start.

**Comment by SuslikV** [comment]
> maybe "Current frame" or "Current frame number"? Or even something like "Preview frame number"...

I will take this into account. This confuses you, thus it will confuse other users. "Current Frame" is sounds good for me.

> ...setRange()

Will be implemented.

> Why a QDoubleSpinBox?..

Total frame numbers in the movie that libopenshot can handle as large as int64_t, so decision was based on this (do you remember the https://github.com/OpenShot/libopenshot/commit/5e1e8ce601952db3160000998e54454cc2da77cd ?). Later, I found that _Export_ dialog are limited on numbers that can be inserted... If max number will be increased - the widget size grows horizontally (probably this is only reason for limitation in _Export_ dialog itself).

*** ? ***
> Try dragging the edge of a clip. The _video_ preview frame changes. It follows the drag position, and previews only the clip being resized (instead of the entire timeline)...

Hmm, probably this is another bug in the system, because when I dragging clip - cursor of the _Timeline_ stays at place and _Video Preview_ shows _Timeline_ (Win7x64). Where should I drag? I'm dragging Clip's edge in the Timeline itself.

> When @N3WWN added Save Frame, he not only made sure to do that, he even _added instructions_ for anyone making similar changes in the future...

Interface shouldn't be based on "px" size. It was temporary workaround to place empty widget (bad practice in UI development).

I can rewrite the UI completely, but in this case the feature will never be accepted. Such code will posses whole project. And I better will develop my own fork rather than wait for 5-6 years when developers will dig into the changes in try to understand how it will affect their code and how to work with their own project from now on.

Anyway, I will try to change the centered layout of the buttons completely. For example, I never used other buttons but play/pause in Video Preview. I do Frames manipulation only by keyboard arrows. Maybe mac users need this buttons, there were reports (https://github.com/OpenShot/openshot-qt/issues/811) that arrows didn't work on mac - but there is no buttons to advance by one frame forward or backward - that are illogical for me... Existing "Fast Forward" and "Rewind" descriptions are confuses me even more - it's digital video editing - what to rewind and fast forward, or for what? Maybe play backward? Then there should be only one button that switches direction of the playback... It is so strange UI, full of items that nobody needs.

**Comment by SuslikV** [comment]
Also I looked again into the button's actions. And turns out that Rewind is never worked OK. After the pause it playbacks the file forward. And this buttons only controls speed of the real-time playback (questionable function itself - there is timeline manipulation over the clip to change the speed and result will be surely exported, but this option is not):
"<<" = "reduce speed or -1.0";
">>" = "increase speed or +1.0".
This can be substituted with the slider or even drop down list with the set of "Preview Speed": -32x -16x -8x -4x -2x -1x 1x 2x 4x 8x 16x 32x.

**Comment by SuslikV** [comment]
Video Preview UI would be like this:

OpenShot new control buttons01

minimum and efficient (at least, I hope for)

**Comment by SuslikV** [comment]
It seems that _libopenshot_ **doesn't  supports** float values for speed right now
https://github.com/OpenShot/libopenshot/blob/6e7ad2316b56f6219e67b83e37ed43b5315544ce/include/Qt/VideoCacheThread.h#L77
Hope, soon it will be possible to set something like 0.2x speed, so QDoubleSpinBox will be used again here (instead of integer controls).

File: src/windows/main_window.py [hunk 1]
```diff
@@ -2689,6 +2717,15 @@ def __init__(self, mode=None):
         # Save settings
         s.save()
 
+        # Timer to apply last Navigation changes with 0.5 sec delay if new one is not come up yet
+        self.navigateToFrame = 1
+        self.navigateToFrame_timer = QTimer()
+        self.navigateToFrame_timer.setSingleShot(True)
+        self.navigateToFrame_timer.setInterval(500)
+        self.navigateToFrame_timer.timeout.connect(self.navigateToFrameTimeout)
+        self.timelines_frame.valueChanged.connect(self.movePlayheadFrames)
```
[the remark this answers is not in the document]
Comment (SuslikV) [reply src/windows/main_window.py:2755]: The `movePlayheadFrames()` was intended to extend timer for as long as user typing at <0.5 sec per character rate. To not navigate to number 4 then to 45 then to 451 if 451 is entered fast enough.

**Comment by SuslikV** [comment]
Icons cache (openshot.qrc) wasn't updated - I don't have environment where to perform required operations. If required, corresponding changes should be added as new commit.

**Comment by SuslikV** [comment]
I bring the "bad practice" (of building UI based on pixels) to new level. Three spacers now. 2 of them is modified when size of the rendering area changes. ^_^

**Comment by SuslikV** [comment]
The worst scenario is when the width of the video rendered area as smaller than 320 px. Symmetry broken:

OpenShot Preview navigation less than 320px width

however, it keeps symmetry for larger sizes (>=320 px):

OpenShot Preview navigation more than 320px width

The pros of the new layout is that the _Playback_ button is almost always visible (width less than 50 px):

OpenShot Preview navigation less than 50px width

The https://github.com/OpenShot/openshot-qt/pull/3346/commits/26199e8790ebb382b767d05e13334e2043660801 actually fixes strange behavior of the navigation buttons. For example, the "rewind" starts playback in forward direction (originally it was intended to control the speed of the playback). Also, it adds "1 frame forward"/"backward" buttons (good to have it in UI too).

**Comment by SuslikV** [comment]
Current _develop_ branch was changed. I will update the code later.

**Comment by SuslikV** [comment]
Updated on top of the current _develop_.

**Comment by jonoomph** [comment]
Closing PR. Trying to maintain simplicity for new users, and not adding more text widgets to the toolbar.

**Pull request closed without merging**